File: src/distance_warning/scripts/distance_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from distance_warning_interfaces.msg import Distance
import random


class DistancePublisher(Node):
    def __init__(self):
        super().__init__('distance_publisher')
        
        self.publisher = self.create_publisher(Distance, 'distance_topic', 10)
        self.timer     = self.create_timer(1.0, self.timer_callback)
        
        self.get_logger().info(f'Distance Publisher initialized, publishing at 1 Hz')

    def timer_callback(self):
        msg = Distance()
        msg.distance = random.uniform(0.1, 1.5)
        
        self.publisher.publish(msg)
        self.get_logger().info(f'Published distance: {msg.distance:.2f} m')


def main(args=None):
    rclpy.init(args=args)
    node = DistancePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()


if __name__ == '__main__':
    main() 

File: src/distance_warning/src/distance_publisher.cpp
#include "distance_warning/distance_publisher.hpp"

DistancePublisher::DistancePublisher()
    : Node("distance_publisher"),
     m_gen(m_rd()),
      m_dist(MIN_DISTANCE, std::nextafter(MAX_DISTANCE, std::numeric_limits<float>::max()))
{
    m_publisher = this->create_publisher<Distance>("distance_topic", 10);
    m_timer     = this->create_wall_timer(
        1s, std::bind(&DistancePublisher::timer_callback, this));
    
    RCLCPP_INFO(this->get_logger(), "distance_publisher initialized, publishing at 1 Hz");
}

void DistancePublisher::timer_callback() {
    auto msg     = Distance();
    msg.distance = m_dist(m_gen);
    
    m_publisher->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Published distance: %.2f m", msg.distance);
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DistancePublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
